audio-echo/app/src/main/java/com/google/sample/echo/MainActivity.java

```diff
@@ -120,11 +133,46 @@ private void queryNativeAudioParameters() {
         nativeSampleFormat ="";    //TODO: find a way to get the native audio format
     }
     private void updateNativeAudioUI() {
-        status_view.setText("nativeSampleRate    = " + nativeSampleRate + "\n" +
-                "nativeSampleBufSize = " + nativeSampleBufSize + "\n" +
-                "nativeSampleFormat  = " + nativeSampleFormat);
+        if(permissionGranted) {
+            status_view.setText("nativeSampleRate    = " + nativeSampleRate + "\n" +
+                    "nativeSampleBufSize = " + nativeSampleBufSize + "\n" +
+                    "nativeSampleFormat  = " + nativeSampleFormat);
+        } else {
+            status_view.setText("Error: Resource Permission not Granted" +
+                                "could not record Audio\n"    +
+                                 "Please restart App and grant the necessary permissions");
+        }
+    }
+    private void requestResourcePermission() {
+        if (ActivityCompat.checkSelfPermission(this, Manifest.permission.RECORD_AUDIO) !=
+                PackageManager.PERMISSION_GRANTED) {
+            ActivityCompat.requestPermissions(this, new String[]{
+                            Manifest.permission.RECORD_AUDIO,
+                            Manifest.permission.MODIFY_AUDIO_SETTINGS,
+                            Manifest.permission.WRITE_EXTERNAL_STORAGE },
```

Comment:
for timing issues, you could not logcat to pring out: it takes too much time. The timing is critical for audio applications, the best way is to log into memory buffer and flush to disk when buffer is full or app closing so writing file is not a factor to audio timing.  I found I could find the timing irregularity when this simple logging so I used it.  Printing out to standout takes long [ I do not have numbers for comparison though]